src/hotspot/share/services/attachListener.cpp

```diff
@@ -472,3 +511,166 @@ void AttachListener::detachall() {
+  }
+}
+
+bool AttachOperation::read_request(RequestReader* reader) {
+  uint ver = reader->read_uint();
+  int buffer_size = 0;
+  // Read conditions:
+  int min_str_count = 0; // expected number of strings in the request
+  int min_read_size = 1; // expected size of the request data (by default 1 symbol for terminating '\0')
+  switch (ver) {
+  case ATTACH_API_V1: // <ver>0<cmd>0<arg>0<arg>0<arg>0
+    // Always contain a command (up to name_length_max symbols)
+    // and arg_count_max(3) arguments (each up to arg_length_max symbols).
+    buffer_size = (name_length_max + 1) + arg_count_max * (arg_length_max + 1);
+    min_str_count = 1 /*name*/ + arg_count_max;
+    break;
+  case ATTACH_API_V2: // <ver>0<size>0<cmd>0<arg>0<arg>0<arg>0
+    if (AttachListener::get_supported_version() < 2) {
+        log_error(attach)("Failed to read request: v2 is unsupported ot disabled");
+        return false;
+    }
+
+    // read size of the data
+    buffer_size = reader->read_uint();
+    if (buffer_size < 0) {
+      return false;
+    }
+    log_debug(attach)("v2 request, data size = %d", buffer_size);
+
+    // Sanity check: max request size is 256K.
+    if (buffer_size > 256 * 1024) {
+      log_error(attach)("Failed to read request: too big");
+      return false;
+    }
+    // Must contain exact 'buffer_size' bytes.
+    min_read_size = buffer_size;
+    break;
+  default:
+    log_error(attach)("Failed to read request: unknown version (%d)", ver);
+    return false;
+  }
+
+  char* buffer = (char*)os::malloc(buffer_size, mtServiceability);
+  int str_count = 0;
+  int off = 0;
+  int left = buffer_size;
+
+  // Read until all (expected) strings or expected bytes have been read, the buffer is full, or EOF.
+  do {
+    int n = reader->read(buffer + off, left);
+    if (n < 0) {
+      os::free(buffer);
+      return false;
+    }
+    if (n == 0) { // EOF
+      break;
+    }
+    if (min_str_count > 0) { // need to count arguments
+      for (int i = 0; i < n; i++) {
+        if (buffer[off + i] == '\0') {
+          str_count++;
+        }
+      }
+    }
+    off += n;
+    left -= n;
+  } while (left > 0 && (off < min_read_size || str_count < min_str_count));
+
+  if (off < min_read_size || str_count < min_str_count) { // unexpected EOF
+    log_error(attach)("Failed to read request: incomplete request");
+    os::free(buffer);
+    return false;
+  }
+  // Request must ends with '\0'.
+  if (buffer[off - 1] != '\0') {
+    log_error(attach)("Failed to read request: not terminated");
+    os::free(buffer);
+    return false;
+  }
+
+  // Parse request.
+  // Command name is the 1st string.
+  set_name(buffer);
+  log_debug(attach)("read request: cmd = %s", buffer);
+
+  // Arguments.
+  char* end = buffer + off;
+  for (char* cur = strchr(buffer, '\0') + 1; cur < end; cur = strchr(cur, '\0') + 1) {
+    log_debug(attach)("read request: arg = %s", cur);
+    append_arg(cur);
+  }
+
+  os::free(buffer);
+
+  return true;
+}
```

Comment:
Nit: This function is too big. I'd suggest to split it to make more readable. For example, the lines 596-648 can be moved to new function which is called by the `AttachOperation::read_request()'.

Reply:
Done